module/zfs/dsl_dataset.c

```diff
@@ -1515,18 +1515,20 @@ dsl_dataset_snapshot_check_impl(dsl_dataset_t *ds, const char *snapname,
 			return (error);
 	}
 
-	error = dsl_dataset_snapshot_reserve_space(ds, tx);
-	if (error != 0)
-		return (error);
-
 	return (0);
 }
 
+struct checknode {
+    list_node_t link;
```

Comment:
It'd be nice to be able to do this using a `list_node_t` embedded in the `dsl_dataset_t`.  Then we wouldn't have to allocate and free these small `checknode` structures.  This should be possible with a little care since currently the sync tasks are run sequentially for a pool.  See `dsl_pool_sync()`.

Reply:
Ok.  Tom also talked about doing this.